### PR TITLE
Introduce API endpoint for retrieving parameters of a template

### PR DESCRIPTION
**Pull request by kritika-singh3**
Issue: #3194

Description:
 - added an endpoint `/go/api/admin/templates/:template_name/parameters` which returns a list of parameters that needs to be defined while using it to create a pipeline

curl:

```curl
  curl "http://go-server-url/go/api/admin/templates/template-name/parameters" \
     -u 'username:password' \
     -H 'Accept: application/vnd.go.cd.v7+json'
```
Response:

```json
{
  "_links" : {
    "self" : {
      "href" : "http://go-server-url/go/api/admin/templates/template-name"
    },
    "doc" : {
      "href" : "https://api.gocd.org/#template-config"
    },
    "find" : {
      "href" : "http://go-server-url/go/api/admin/templates/:template_name"
    }
  },
  "name" : "template-name",
  "parameters" : [ "resources", "command" ]
}
```

### CODE REVIEW

File: api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java [hunk 1]
```diff
@@ -170,6 +173,16 @@ public String show(Request req, Response res) throws IOException {
         }
     }
 
+    public String showParameters(Request req, Response res) throws IOException {
+        PipelineTemplateConfig templateConfig = fetchEntityFromConfig(req.params("template_name"));
+        if (isGetOrHeadRequestFresh(req, templateConfig)) {
```
Comment (GaneshSPatil) [inline api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java]: Here, we're checking for the ETag generated based on the entire template config, whereas, this API is only for the template parameters.

Imagine there is a change in the template config, but the template parameters hasn't changed. Where, the API would respond `412` even if there is no change in the response of the API.

---

Here are the options we've:
1. Compute ETag only for template parameters.
2. No ETag for this API?? we always send `200`?
3. Keep the implementation as is (i.e. rely on entire template config for etag).

@kritika-singh3 @ketan @arvindsv @maheshp @bdpiparva -- Thoughts??

Comment (bdpiparva) [reply api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java]: IMO, Etag should be generated based on the payload or entity being serialised. Generating based on entire config does not make sense.

Comment (kritika-singh3) [reply api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java]: I was of the opinion that the parameters don't exist out of template. They can change with any change in the template config.
Scenarios I considered:
1. Parameters being added/removed.
2. The place where the parameters are utilized being changed.

So, decided to do a Etag check on the whole config. If anything, I would like to go with Option 2 as we are fetching the entire template config either way.

Comment (bdpiparva) [reply api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java]: @kritika-singh3 - Agreed, that parameters does not exist outside of  template config or can't be updated directly though API. I am leaning towards option 1 as the value of parameters can remain same when other parts of the template config is changed which defeats the purpose of having etag.

Comment (kritika-singh3) [reply api/api-template-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/templateconfig/TemplateConfigControllerV7.java]: @GaneshSPatil @bdpiparva I have updated the API to calculate eTag only for the parameters and using the same to send either `304` or `200`

**Comment by arvindsv** [comment]
Verified on 20.2.0 (11308-05dc4d14977d4bf10d5f5c9e9d6f88817df1ec1e) with:

### Scenario 1: Basic check:

Config:

```xml
<templates>
  <pipeline name="t1">
    <stage name="s2">
      <jobs>
        <job name="j2">
          <tasks>
            <exec command="echo">
              <arg>#{UPSTREAM}</arg>
            </exec>
          </tasks>
        </job>
      </jobs>
    </stage>
  </pipeline>
</templates>
```

Output:

```
$ curl -v "http://localhost:8153/go/api/admin/templates/t1/parameters" -H 'Accept: application/vnd.go.cd.v7+json'
< HTTP/1.1 200 OK
< ETag: "40d570cbc3281b0855b044c715218648"

{
  "_links" : {
    "self" : {
      "href" : "http://localhost:8153/go/api/admin/templates/t1"
    },
    "doc" : {
      "href" : "https://api.gocd.org/20.2.0/#template-config"
    },
    "find" : {
      "href" : "http://localhost:8153/go/api/admin/templates/:template_name"
    }
  },
  "name" : "t1",
  "parameters" : [ "UPSTREAM" ]
}
```

### Scenario 2: Changed template, but not parameter:


Config:

```xml
<templates>
  <pipeline name="t1">
    <stage name="s2">
      <jobs>
        <job name="j2">
          <tasks>
            <exec command="ls">
              <arg>#{UPSTREAM}</arg>
            </exec>
          </tasks>
        </job>
      </jobs>
    </stage>
  </pipeline>
</templates>
```

Output:

```
$ curl -v "http://localhost:8153/go/api/admin/templates/t1/parameters" -H 'Accept: application/vnd.go.cd.v7+json'
< HTTP/1.1 200 OK
< ETag: "40d570cbc3281b0855b044c715218648"

{
  "_links" : {
    "self" : {
      "href" : "http://localhost:8153/go/api/admin/templates/t1"
    },
    "doc" : {
      "href" : "https://api.gocd.org/20.2.0/#template-config"
    },
    "find" : {
      "href" : "http://localhost:8153/go/api/admin/templates/:template_name"
    }
  },
  "name" : "t1",
  "parameters" : [ "UPSTREAM" ]
}
```

ETag remained the same.

### Scenario 3: Changed template and parameter:

Config:

```xml
<templates>
  <pipeline name="t1">
    <stage name="s2">
      <jobs>
        <job name="j2">
          <tasks>
            <exec command="ls">
              <arg>#{UPSTREAMX}</arg>
            </exec>
          </tasks>
        </job>
      </jobs>
    </stage>
  </pipeline>
</templates>
```

Output:

```
$ curl -v "http://localhost:8153/go/api/admin/templates/t1/parameters" -H 'Accept: application/vnd.go.cd.v7+json'
< HTTP/1.1 200 OK
< ETag: "0c88c98f6e93b89d0256eb97b26d3326"

{
  "_links" : {
    "self" : {
      "href" : "http://localhost:8153/go/api/admin/templates/t1"
    },
    "doc" : {
      "href" : "https://api.gocd.org/20.2.0/#template-config"
    },
    "find" : {
      "href" : "http://localhost:8153/go/api/admin/templates/:template_name"
    }
  },
  "name" : "t1",
  "parameters" : [ "UPSTREAMX" ]
}
```

ETag did change.

### Scenario 4: Removed all parameters:

Config:

```xml
<templates>
  <pipeline name="t1">
    <stage name="s2">
      <jobs>
        <job name="j2">
          <tasks>
            <exec command="ls">
              <arg>abc</arg>
            </exec>
          </tasks>
        </job>
      </jobs>
    </stage>
  </pipeline>
</templates>
```

Output:

```
$ curl -v "http://localhost:8153/go/api/admin/templates/t1/parameters" -H 'Accept: application/vnd.go.cd.v7+json'
< HTTP/1.1 200 OK
< ETag: "d41d8cd98f00b204e9800998ecf8427e"

{
  "_links" : {
    "self" : {
      "href" : "http://localhost:8153/go/api/admin/templates/t1"
    },
    "doc" : {
      "href" : "https://api.gocd.org/20.2.0/#template-config"
    },
    "find" : {
      "href" : "http://localhost:8153/go/api/admin/templates/:template_name"
    }
  },
  "name" : "t1",
  "parameters" : [ ]
}
```

ETag changed and got an empty list of parameters.


### Other observations:

1. If there are multiple parameters in the template, then the order does not affect the ETag.

2. Fails with a 404 (as expected) if you use the header 'Accept: application/vnd.go.cd.v6+json' (v7 is the right version).

3. Tests [here](https://github.com/gocd/gocd/pull/7670/files#diff-939d48a47d7d3c735af7dce816050c2cR27) seem reasonable and no change in [this functional test](https://github.com/gocd/ruby-functional-tests/blob/647d4cf9927a4c3c7e123f0b5b1412b8e9123e69/specs/TemplatesAPI.spec) is needed.

**Pull request merged**